packages/react-instantsearch-hooks/src/__tests__/InstantSearchSSRProvider.test.tsx

```diff
@@ -0,0 +1,275 @@
+}
+
+function Hits() {
+  const { hits } = useHits();
+
+  if (hits.length === 0) {
+    return null;
+  }
+
+  return (
+    <ol>
+      {hits.map((hit) => (
+        <li key={hit.objectID}>{hit.objectID}</li>
+      ))}
+    </ol>
+  );
+}
+
+describe('InstantSearchSSRProvider', () => {
+  test('provides initialResults to InstantSearch', async () => {
+    const searchClient = createSearchClient();
+    const initialResults = {
+      indexName: {
+        state: {},
+        results: [
+          {
+            exhaustiveFacetsCount: true,
+            exhaustiveNbHits: true,
+            hits: [{ objectID: '1' }, { objectID: '2' }, { objectID: '3' }],
+            hitsPerPage: 20,
+            index: 'indexName',
+            nbHits: 0,
+            nbPages: 0,
+            page: 0,
+            params: '',
+            processingTimeMS: 0,
+            query: '',
+          },
+        ],
+      },
+    };
+
+    function App() {
+      return (
+        <InstantSearchSSRProvider initialResults={initialResults}>
+          <InstantSearch searchClient={searchClient} indexName="indexName">
+            <Hits />
+          </InstantSearch>
+        </InstantSearchSSRProvider>
+      );
+    }
+
+    render(<App />);
+
+    await waitFor(() => {
+      expect(screen.getByRole('list')).toMatchInlineSnapshot(`
+<ol>
+  <li>
+    1
+  </li>
+  <li>
+    2
+  </li>
+  <li>
+    3
+  </li>
+</ol>
+`);
+    });
+  });
+
+  test('renders UI state with initialUiState', async () => {
+    const searchClient = createSearchClient();
+    const initialResults = {
+      indexName: {
+        state: { query: 'iphone' },
+        results: [
+          {
+            exhaustiveFacetsCount: true,
+            exhaustiveNbHits: true,
+            hits: [{ objectID: '1' }, { objectID: '2' }, { objectID: '3' }],
+            hitsPerPage: 20,
+            index: 'indexName',
+            nbHits: 0,
+            nbPages: 0,
+            page: 0,
+            params: '',
+            processingTimeMS: 0,
+            query: '',
+          },
+        ],
+      },
+    };
+
+    function App() {
+      return (
+        <InstantSearchSSRProvider initialResults={initialResults}>
+          <InstantSearch
+            searchClient={searchClient}
+            indexName="indexName"
+            initialUiState={{
+              indexName: {
+                query: 'iphone',
+              },
+            }}
+          >
+            <SearchBox />
+            <Hits />
+          </InstantSearch>
+        </InstantSearchSSRProvider>
+      );
+    }
+
+    render(<App />);
+
+    await waitFor(() => {
+      expect(screen.getByRole('search')).toMatchInlineSnapshot(`
+<form
+  role="search"
+>
+  <input
+    value="iphone"
+  />
+</form>
+`);
+    });
+  });
+
+  test('with router provides initial UI state to InstantSearch', async () => {
+    const searchClient = createSearchClient();
+    const initialResults = {
+      indexName: {
+        state: { query: 'iphone' },
```

Comment:
IIUC this state should be computed by `getServerState` so it's unlikely that the `query` in `initialResults` be different from the location one (since the former would come from the latter), correct?

What happens if they differ?

Reply:
Correct. In a real use case, `results[0].query` would have been "iphone" too. This test makes sure that the `<SearchBox>` render state is read by the provided helper state though, not from the result. So I think it makes sense to have it inconsistent here?

Reply:
It does, yes.

By the helper state you mean l. 153 right? It's still unclear to me when I read the test from where "iphone" is picked up, here or `getLocation`.

Reply:
The test name was confusing indeed. I renamed the two tests in fa7f06c327835ac67a1b2a808f96b2c0cccc7d55.

What these tests do is that they make sure that if you pass `initialUiState` or `routing`, the state coming from the initial results is the one used during render.